### DISCUSSION TITLE
Can I Push messages from IOT devices using TCP?

### DISCUSSION THREAD

**Discussion by AamerHejazi** [emqx/MQTTX#1520] [Q&A] [answered]
Dears,
I hope you are doing well,

Can I push data directly from any IOT device without mentioning the TOPIC name using TCP protocol (without the Flespi tool)?

Please advise me and suggest if that is possible.

Best Regards,
Amer Hijazi,

**Comment by ysfscream** [comment]
Hi @AamerHejazi ,

I think it's technically feasible regarding your question about IoT data transmission via TCP without specifying a topic. While topics are a core part of IoT protocols like MQTT, TCP operates at a lower level and handles the raw data transfer between devices and servers. You can transmit data directly over TCP, but ensure the receiving server is set up to interpret the data stream correctly. However, note that bypassing higher-level protocols like MQTT may result in losing functionalities such as easy message routing and quality of service. If you have specific constraints or use cases, please provide more details for a more targeted suggestion.

Best regards,
Shifan

**Reply by AamerHejazi** [reply]
Thank you for your support and answer @ysfscream

**Comment by AamerHejazi** [comment]
Dear @ysfscream,
Thank you for your response.

I am currently working with an old system that accommodates various types of tracker devices. These devices connect to our server via TCP, and our backend is set up to listen to this port, converting the incoming hexadecimal data into JSON format.

The challenge arises with devices from Teltonika and other manufacturers that do not support the MQTT protocol. I have been in contact with Teltonika, and while they do offer some devices compatible with MQTT, these are limited to functioning specifically with Google Cloud or Azure platforms.

Given these constraints, I am exploring potential solutions or workarounds. Would you have any suggestions or recommendations on how to integrate these non-MQTT devices into the EMQX broker and take the most benefits from it? Any advice or alternative approaches you can suggest would be greatly appreciated.

Thank you for considering my query. I look forward to any insights you might have.

Best Regards,
Amer Hijazi

**Reply by ysfscream** [reply, marked as answer]
Regarding the Teltonika and other tracker devices that do not support the MQTT protocol, one possible solution could be to create a middleware service. This service would act as a receiving point for TCP connections and data, converting the received hexadecimal data into JSON format and then forwarding it to the EMQX broker using the MQTT protocol.

This middleware service could be a flexible solution for devices that inherently do not support MQTT. It would handle the protocol translation, enabling these devices to communicate indirectly with the EMQX broker. This way, you can take advantage of the many benefits of EMQX, such as scalability and robust message-handling capabilities.

Another feasible solution is to utilize gateway protocol plugins within EMQX. These plugins enable the EMQX broker to accept connections over various protocols, including TCP. Please check if a plugin suits your case, or you can explore the possibility of developing a new one. For instance, EMQX-supported protocol plugins might include COAP, STOMP, etc., which can effectively connect devices using different protocols.

I understand these solutions require additional development work, but they offer viable ways to effectively integrate your non-MQTT devices with the EMQX broker.

Ref: https://docs.emqx.com/en/enterprise/v4.4/modules/gateway.html

<img width="782" alt="image" src="https://github.com/emqx/MQTTX/assets/21299158/e97d45fa-f407-4d79-bb0b-edf80697aa1b">

**Reply by AamerHejazi** [reply]
Thank you for your support and answer @ysfscream
